### PR TITLE
Improve Unicode perf.

### PR DESCRIPTION
**Pull request by BillyONeal**
In https://github.com/microsoft/vcpkg-tool/pull/1316 I noticed that callers expect `pointer_to_current` to be cheap, and we were doubling the number of branches by converting to enums first.

For `vcpkg format-manifest --all` this change results in:

	After	Before
	1.3853984	1.4387856
	1.3930014	1.5002737
	1.4488841	1.4187639
	1.3823215	1.4723308
	1.4366255	1.4213668
	1.3880373	1.4606811

Average	1.405711367	1.45203365
StdDev	0.026624509	0.028973257

(@ras0219-msft edit: -3.2% faster)

### CODE REVIEW

File: src/vcpkg/base/unicode.cpp [hunk 1]
```diff
@@ -3,176 +3,218 @@
...
-        return {it, utf8_errc::NoError};
+        out = end_of_file;
+        first = last;
+        return utf8_errc::InvalidCodeUnit;
     }
 
     bool utf8_is_valid_string(const char* first, const char* last) noexcept
     {
-        utf8_errc err = utf8_errc::NoError;
-        for (auto dec = Utf8Decoder(first, last); dec != dec.end(); err = dec.next())
```
Comment (BillyONeal) [inline src/vcpkg/base/unicode.cpp:173]: Stealth bug fix: the previous implementation would crash if the first code point was invalid.

File: include/vcpkg/base/unicode.h [hunk 2]
```diff
@@ -29,18 +19,12 @@ namespace vcpkg::Unicode
         UnexpectedEof = 6,
     };
 
-    Utf8CodeUnitKind utf8_code_unit_kind(unsigned char code_unit) noexcept;
-
-    int utf8_code_unit_count(char code_unit) noexcept;
-
     int utf8_encode_code_point(char (&array)[4], char32_t code_point) noexcept;
 
     // returns {after-current-code-point, error},
```
Comment (BillyONeal) [inline include/vcpkg/base/unicode.h:24]: Wrong

File: src/vcpkg/base/unicode.cpp [hunk 3]
```diff
@@ -222,28 +222,34 @@ namespace vcpkg::Unicode
...
+            pointer_to_current_ = last;
+            next_ = last;
             return utf8_errc::PairedSurrogates;
         }
 
-        next_ = new_next.first;
         current_ = code_point;
+        pointer_to_current_ = old_next;
         return utf8_errc::NoError;
     }
```
Comment (ras0219-msft) [inline src/vcpkg/base/unicode.cpp:255]: Can we get rid of `operator++`? Or is it in use somewhere else?

**Review by ras0219-msft — APPROVED** [review]

File: src/vcpkg/base/unicode.cpp (hunk 3 above)
Comment (BillyONeal) [reply src/vcpkg/base/unicode.cpp:255]: I'm pretty sure lots of places use this as an iterator still

File: src/vcpkg/base/unicode.cpp [hunk 4]
```diff
@@ -3,176 +3,187 @@
...
-                msg::format(msgInvalidCodePoint).append_raw(fmt::format("({:x})", static_cast<uint32_t>(code_point))));
-        }
+
+        vcpkg::Checks::msg_exit_with_message(
+            VCPKG_LINE_INFO,
+            msg::format(msgInvalidCodePoint).append_raw(fmt::format("({:x})", static_cast<uint32_t>(code_point))));
     }
 
-    int utf8_encode_code_point(char (&array)[4], char32_t code_point) noexcept
+    static bool bad_trailing(unsigned char code_unit, utf8_errc& err) noexcept
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/base/unicode.cpp]: Fixed!

File: src/vcpkg/base/unicode.cpp (hunk 3 above)
Comment (BillyONeal) [reply src/vcpkg/base/unicode.cpp:253]: Yeah this is used all over.

**Comment by BillyONeal** [comment]
After applying Robert's suggestion I now get:

image

if I throw out the outlier 4.3 second one:

image

(These are on the Zypherus G14 / Ryzen 9 7940HS)

**Comment by BillyONeal** [comment]
Reran on the 7950X and the values are closer. Don't know what to make of that :shrug: 
image

**Comment by BillyONeal** [comment]
I ran more trials and maybe on the 7950X it makes no difference :/. I still like this change though

image

**Comment by BillyONeal** [comment]
OK I give up trying to characterize it :P

image

**Pull request merged**